[GStreamer] MediaPlayerPrivateGStreamer::sourceSetup() can be called from a non-main thread in some cases

philn opened the pull request:
#### d39df7316b4a332ead8b4b6191303648c4df6042
<pre>
[GStreamer] MediaPlayerPrivateGStreamer::sourceSetup() can be called from a non-main thread in some cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=258132">https://bugs.webkit.org/show_bug.cgi?id=258132</a>

Reviewed by Xabier Rodriguez-Calvar.

The crash was happening due to usage of the WeakPtr m_player in non-main non-GC thread. The proposed
solution is to pass a reference to the PlatformMediaResource loader of the player to the HTTP source
element, thus avoiding usage of m_player in non-main non-GC thread.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::sourceSetup):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcSetResourceLoader):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/265298@main">https://commits.webkit.org/265298@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7f0a580323caeff17f5dc96b9a7839d03153015

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10952 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12917 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10633 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12981 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12487 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9451 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16719 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9603 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12859 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8146 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9200 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2520 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13456 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -925,11 +925,11 @@ static void webKitWebSrcUriHandlerInit(gpointer gIface, gpointer)
 }
 
-void webKitWebSrcSetResourceLoader(WebKitWebSrc* src, RefPtr<WebCore::PlatformMediaResourceLoader>&& loader)
+void webKitWebSrcSetResourceLoader(WebKitWebSrc* src, RefPtr<WebCore::PlatformMediaResourceLoader> loader)
```
Shouldn't this be:

```suggestion
void webKitWebSrcSetResourceLoader(WebKitWebSrc* src, WebCore::PlatformMediaResourceLoader& loader)
```

? Passing a RefPtr by value is probably not desired?

philn replied:
Why not?

mcatanzaro replied:
You'll have an unnecessary ref and unref

philn replied:
I see no = operator in RefPtr for references. Passing a `const RefPtr<>&` would work though, I think?

mcatanzaro replied:
Sorry, I was incorrect to suggest using a raw reference, since this is a RefPtr, not a Ref. Should use a raw pointer instead.

There are two cases:

 * You want to move ownership. Use `Ref<Foo>&&` or `RefPtr<Foo>&&`
 * You want the caller to retain ownership. Use `Foo&` or `Foo*` to indicate ownership does not move.

The latter case applies here (even though you do eventually take ownership down below).

philn committed:
[GStreamer] MediaPlayerPrivateGStreamer::sourceSetup() can be called from a non-main thread in some cases
https://bugs.webkit.org/show_bug.cgi?id=258132

Reviewed by Xabier Rodriguez-Calvar.

The crash was happening due to usage of the WeakPtr m_player in non-main non-GC thread. The proposed
solution is to pass a reference to the PlatformMediaResource loader of the player to the HTTP source
element, thus avoiding usage of m_player in non-main non-GC thread.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::sourceSetup):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcSetResourceLoader):
* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.h:

Canonical link: https://commits.webkit.org/265298@main
